File: android/src/com/android/tools/idea/avdmanager/ChooseSystemImageStep.java
/*
 * Copyright (C) 2015 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.android.tools.idea.avdmanager;

import com.android.sdklib.repository.targets.SystemImage;
import com.android.tools.idea.ui.properties.core.ObservableBool;
import com.android.tools.idea.ui.validation.Validator;
import com.android.tools.idea.ui.validation.ValidatorPanel;
import com.android.tools.idea.ui.wizard.StudioWizardStepPanel;
import com.android.tools.idea.wizard.model.ModelWizard;
import com.android.tools.idea.wizard.model.ModelWizardStep;
import com.android.tools.swing.util.FormScalingUtil;
import com.google.common.base.Optional;
import com.intellij.openapi.project.Project;
import com.intellij.util.Consumer;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import javax.swing.*;

/**
 * Wizard step for selecting a {@link SystemImage} from the installed images in the SDK.
 */
public class ChooseSystemImageStep extends ModelWizardStep<AvdOptionsModel> {
  @Nullable private final Project myProject;

  private JPanel myPanel;
  private ChooseSystemImagePanel myChooseImagePanel;
  private StudioWizardStepPanel myStudioWizardStepPanel;
  private ValidatorPanel myValidatorPanel;

  public ChooseSystemImageStep(@NotNull AvdOptionsModel model, @Nullable Project project) {
    super(model, "System Image");
    myProject = project;
    myValidatorPanel = new ValidatorPanel(this, myPanel);
    myStudioWizardStepPanel = new StudioWizardStepPanel(myValidatorPanel, "Select a system image");
    FormScalingUtil.scaleComponentTree(this.getClass(), myStudioWizardStepPanel);

    myChooseImagePanel.addSystemImageListener(new Consumer<SystemImageDescription>() {
      @Override
      public void consume(SystemImageDescription systemImage) {
        getModel().systemImage().setNullableValue(systemImage);
      }
    });
  }

  @Override
  protected void onWizardStarting(@NotNull ModelWizard.Facade wizard) {
    myValidatorPanel.registerValidator(getModel().systemImage(), new Validator<Optional<SystemImageDescription>>() {
      @NotNull
      @Override
      public Result validate(@NotNull Optional<SystemImageDescription> value) {
        return (value.isPresent())
               ? Result.OK
               : new Validator.Result(Severity.ERROR, "A system image must be selected to continue.");
      }
    });
  }

  @Override
  protected void onEntering() {
    myChooseImagePanel.setDevice(getModel().device().getValueOrNull());
  }

  @NotNull
  @Override
  protected ObservableBool canGoForward() {
    return myValidatorPanel.hasErrors().not();
  }

  @NotNull
  @Override
  protected JComponent getComponent() {
    return myStudioWizardStepPanel;
  }

  private void createUIComponents() {
    myChooseImagePanel =
      new ChooseSystemImagePanel(myProject, getModel().device().getValueOrNull(), getModel().systemImage().getValueOrNull());
  }
}

File: android/src/com/android/tools/idea/deviceManager/avdmanager/ChooseSystemImageStep.kt
/*
 * Copyright (C) 2015 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.android.tools.idea.deviceManager.avdmanager

import com.android.tools.adtui.util.FormScalingUtil
import com.android.tools.adtui.validation.Validator
import com.android.tools.adtui.validation.ValidatorPanel
import com.android.tools.idea.observable.core.ObservableBool
import com.android.tools.idea.ui.wizard.deprecated.StudioWizardStepPanel
import com.android.tools.idea.wizard.model.ModelWizard
import com.android.tools.idea.wizard.model.ModelWizardStep
import com.intellij.openapi.project.Project
import com.intellij.ui.layout.LCFlags
import com.intellij.ui.layout.panel
import java.util.Optional
import javax.swing.JComponent

/**
 * Wizard step for selecting a [SystemImage] from the installed images in the SDK.
 */
class ChooseSystemImageStep(model: AvdOptionsModel, project: Project?) : ModelWizardStep<AvdOptionsModel>(model, "System Image") {
  private var chooseImagePanel = ChooseSystemImagePanel(project, model.device().valueOrNull, model.systemImage().valueOrNull).apply {
    addSystemImageListener { model.systemImage().setNullableValue(it) }
  }
  private val panel = panel(LCFlags.fill) {
    row {
      chooseImagePanel(grow)
    }
  }
  private val validatorPanel: ValidatorPanel = ValidatorPanel(this, panel)
  private val studioWizardStepPanel: StudioWizardStepPanel = StudioWizardStepPanel(validatorPanel, "Select a system image")

  init {
    FormScalingUtil.scaleComponentTree(this.javaClass, studioWizardStepPanel)
  }

  override fun onWizardStarting(wizard: ModelWizard.Facade) {
    validatorPanel.registerValidator(
      model.systemImage(),
      object : Validator<Optional<SystemImageDescription>> {
        override fun validate(value: Optional<SystemImageDescription>): Validator.Result =
          if (value.isPresent) Validator.Result.OK
          else Validator.Result(Validator.Severity.ERROR, "A system image must be selected to continue.")
      }
    )
  }

  override fun onEntering() {
    chooseImagePanel.setDevice(model.device().valueOrNull)
  }

  override fun canGoForward(): ObservableBool = validatorPanel.hasErrors().not()

  override fun getComponent(): JComponent = studioWizardStepPanel
}